Applications/Spire/Source/Ui/TextBox.cpp

```diff
@@ -275,3 +357,13 @@ void TextBox::update_display_text() {
     m_line_edit->setText(m_current);
   }
 }
+
+void TextBox::update_placeholder_text() {
+  if(is_placeholder_shown()) {
+    m_placeholder->setText(get_elided_text(m_placeholder->fontMetrics(),
```

Comment:
Here the font metrics are obtained from m_placeholder, but the elided text is calculated by m_line_edit in the get_elided_text function. The elided text therefore is got from the parameters of m_placeholder and m_line_edit, which are the font metrics of m_placehoder and the text area of the m_line_edit. 
The QLineEdit and QLabel actually have different ways to compute the text area, so I used different code to get elided text in the previous version.